dataset/src/main/scala/frameless/TypedEncoder.scala

```diff
@@ -296,30 +343,119 @@ object TypedEncoder {
     }
   }
 
+  implicit def arrayEncoder[A : ClassTag](
+    implicit
+    underlying: TypedEncoder[A]
+  ): TypedEncoder[Array[A]] = new CollectionEncoder[Array, A]() {
+    val nullable: Boolean = false
+
+    val sourceDataType: DataType = FramelessInternals.objectTypeFor[Array[A]](classTag)
+
+    val targetDataType: DataType = DataTypes.createArrayType(underlying.targetDataType)
+
+    def constructorFor(path: Expression): Expression = arrayData(path)
+
+    def extractorFor(path: Expression): Expression = {
+      // if source `path` is already native for Spark, no need to `map`
+      if (ScalaReflection.isNativeType(underlying.sourceDataType)) {
+        NewInstance(
+          classOf[GenericArrayData],
+          path :: Nil,
+          dataType = ArrayType(underlying.targetDataType, underlying.nullable)
+        )
+      } else {
+        MapObjects(underlying.extractorFor, path, underlying.sourceDataType)
+      }
+    }
+  }
+
+  implicit def mapEncoder[A : ClassTag, B : ClassTag](
+    implicit
+    encodeA: TypedEncoder[A],
+    encodeB: TypedEncoder[B]
+  ): TypedEncoder[Map[A, B]] = new TypedEncoder[Map[A, B]] {
+    val nullable: Boolean = false
+    val sourceDataType = FramelessInternals.objectTypeFor[Map[A, B]]
+    val targetDataType = MapType(encodeA.targetDataType, encodeB.targetDataType, encodeB.nullable)
+
+    // needed to construct Map using `ArrayBasedMapData.toScalaMap` (avoids an extra MapObjects)
+    private val arrayA = arrayEncoder[A]
```

Comment:
As I know, code generation in Spark happens only once, then Spark generates Java code and executes it, this code gets executed only on driver node before actual job gets executed, so there is no need to do any kind of optimizations.